hilti/toolchain/include/ast/types/name.h

```diff
@@ -19,13 +17,16 @@ class Name : public UnqualifiedType {
     bool isBuiltIn() const { return _builtin; }
 
     // resolves recursively
-    UnqualifiedType* resolvedType() const {
+    UnqualifiedType* resolvedType(size_t recursion_depth = 0) const {
         if ( ! _resolved_type_index )
             return nullptr;
 
+        if ( recursion_depth > 1000 )
+            return nullptr;
+
         auto t = context()->lookup(_resolved_type_index);
         if ( auto n = t->tryAs<type::Name>() )
-            return n->resolvedType();
+            return n->resolvedType(recursion_depth + 1);
         else
             return t;
     }
```

Comment:
Is there a reason we wouldn't want to use a cycle detector here? It seems that would help avoid the arbitrary recursion limit. Something like e.g.,

```suggestion
    UnqualifiedType* resolvedType(node::CycleDetector* cd = nullptr) const {
        if ( ! _resolved_type_index )
            return nullptr;

        auto t = context()->lookup(_resolved_type_index);
        assert(t);

        if ( auto n = t->tryAs<type::Name>() ) {
            if ( cd && cd->haveSeen(n) )
                return nullptr;

            if ( ! cd ) {
                node::CycleDetector _cd;
                _cd.recordSeen(n);
                return n->resolvedType(&_cd);
            }

            cd->recordSeen(n);
            return n->resolvedType(cd);
        }
        else
            return t;
    }
```

Reply:
I actually prefer the recursion limit. This particular problem comes up with just deeply nested names. For example, I wrote a little script that makes a bunch of types which refer to each other (and the last one is `uint8`):

```
# make-test.rb
f = File.new('deeplynested.spicy', 'w')
f.write("module Test;\n")
num_iterations = 10000
(0..num_iterations).each do |i|
  f.write("type Data%d = Data%d;\n" % [i, i + 1])
end

f.write("type Data%d = uint8;\n" % [num_iterations + 1])

f.close()
```

So running `spicyc` on the generated `deeplynested.spicy` file will segfault in main but not in this branch (though the errors are a little annoying). I figured killing two birds with one stone would be better than keeping around an obvious segfault, even if it is unlikely.

Then it's slightly less overhead to increment a number than add a value to an unordered set and check it, even if it's inconsequential it could theoretically add up

Though, the random `recursion_depth > 1000` limit is pretty rough.

Reply:
Ah I see, that make sense and is indeed unrelated to cycle. Could you add a test which generates such a long chain of type aliases and then captures the new `spicyc` failure so we have it in a baseline?

Reply:
[Added](https://github.com/zeek/spicy/blob/0b1f3214559c7eec894b8d3f906fed86577f2cd4/tests/hilti/types/id/recursion-limit.sh) - though it's a bit of a weird test :)

Reply:
I'll hold off on this for the day, but the CI jobs for this were taking a lot longer than normal, I cancelled them and the `id.recursion-limit` test wasn't run on them (well, I suspect was currently running for a very, very long time). The CI jobs that passed (like centos stream 9) had that test looking proper.

I'm not too sure about that test :/

Reply:
Oops, yeah that looks good, my bad